### ISSUE TITLE
gvisor: panic: Incrementing non-positive count 0xc000533180 on stack.PacketBuffer

### ISSUE THREAD

**Issue by DentonGentry** [tailscale/tailscale#4144]
### What is the issue?

From email report:

we compiled 1.22.1 and are experiencing problems on two of our macOS machines. tailscaled crashes directly after startup.

The ports 6883, 6885 are used by a local program which is running in the background before tailscaled is started.

We do see this error on our 2 Intel macOS systems. Arm macOS does not show the error.

We had a similar problem with 1.22.0. Downgrading on one of the intel machines to 1.20.4 solved the problem.

```
2022/03/10 15:08:59 logtail started
2022/03/10 15:08:59 Program starting: v1.22.1-tc8fb4f8c7, Go 1.17.8-tsdce70b6: []string{"/usr/local/bin/tailscaled", "-state=/var/db/tailscale/tailscaled.state", "-tun", "utun9"}
2022/03/10 15:08:59 LogID: 459c4a8ac464c58028859fb0d0bb79ee88adf8db6a27e083c7a7d63242103f9e
2022/03/10 15:08:59 logpolicy: using system state directory "/Library/Tailscale"
2022/03/10 15:08:59 wgengine.NewUserspaceEngine(tun "utun9") ...
2022/03/10 15:08:59 dns: using dns.noopManager
2022/03/10 15:08:59 link state: interfaces.State{defaultRoute=en13 ifs={en13:[10.0.5.6/24]} v4=true v6=false}
2022/03/10 15:08:59 magicsock: disco key = d:f376519d5ab5ccd3
2022/03/10 15:08:59 Creating wireguard device...
2022/03/10 15:08:59 Bringing wireguard device up...
2022/03/10 15:08:59 Bringing router up...
2022/03/10 15:08:59 external route: up
2022/03/10 15:08:59 Clearing router settings...
2022/03/10 15:08:59 Starting link monitor...
2022/03/10 15:08:59 Engine created.
2022/03/10 15:09:01 Start
2022/03/10 15:09:01 using backend prefs for "_daemon": Prefs{ra=true dns=true want=true routes=[0.0.0.0/0 ::/0 10.0.5.0/24] snat=true Persist{lm=, o=, n=[jR0DB] u="REDACTED"}}
2022/03/10 15:09:01 Backend: logs: be:REDACTED fe:
2022/03/10 15:09:01 control: client.Login(false, 0)
2022/03/10 15:09:01 control: doLogin(regen=false, hasUrl=false)
2022/03/10 15:09:01 control: control server key [nlFWp] from https://controlplane.tailscale.com/
2022/03/10 15:09:01 control: RegisterReq: onode= node=[jR0DB] fup=false
2022/03/10 15:09:02 control: RegisterReq: got response; nodeKeyExpired=false, machineAuthorized=true; authURL=false
2022/03/10 15:09:02 active login: REDACTED
2022/03/10 15:09:02 Switching ipn state NoState -> Starting (WantRunning=true, nm=true)
2022/03/10 15:09:02 health("overall"): error: state=Starting, wantRunning=true
2022/03/10 15:09:02 magicsock: SetPrivateKey called (init)
2022/03/10 15:09:02 wgengine: Reconfig: configuring userspace wireguard config (with 1/10 peers)
2022/03/10 15:09:02 wgengine: Reconfig: configuring router
2022/03/10 15:09:02 wgengine: Reconfig: configuring DNS
2022/03/10 15:09:02 dns: Set: {DefaultResolvers:[] Routes:{} SearchDomains:[] Hosts:11}
2022/03/10 15:09:02 dns: Resolvercfg: {Routes:{} Hosts:11 LocalDomains:[]}
2022/03/10 15:09:02 dns: OScfg: {Nameservers:[] SearchDomains:[] MatchDomains:[]}
2022/03/10 15:09:02 peerapi: serving on http://100.x.y.z:35528/
2022/03/10 15:09:02 peerapi: serving on http://[fd7a:115c:a1e0:ab12:q:r:s:t]:35528/
2022/03/10 15:09:02 magicsock: home is now derp-1 (nyc)
2022/03/10 15:09:02 magicsock: adding connection to derp-1 for home-keep-alive
2022/03/10 15:09:02 control: NetInfo: NetInfo{varies=false hairpin=false ipv6=false udp=true derp=#1 portmap= link=""}
2022/03/10 15:09:02 magicsock: 1 active derp conns: derp-1=cr0s,wr0s
2022/03/10 15:09:02 derphttp.Client.Connect: connecting to derp-1 (nyc)
2022/03/10 15:09:02 Switching ipn state Starting -> Running (WantRunning=true, nm=true)
2022/03/10 15:09:02 magicsock: endpoints changed: a.b.c.d:57147 (stun), 10.0.5.6:57147 (local)
2022/03/10 15:09:02 magicsock: adding connection to derp-4 for [/+7hj]
2022/03/10 15:09:02 magicsock: 2 active derp conns: derp-1=cr83ms,wr83ms derp-4=cr0s,wr0s
2022/03/10 15:09:02 derphttp.Client.Recv: connecting to derp-4 (fra)
2022/03/10 15:09:03 magicsock: derp-1 connected; connGen=1
2022/03/10 15:09:03 health("overall"): ok
2022/03/10 15:09:03 magicsock: derp-4 connected; connGen=1
2022/03/10 15:09:03 wgengine: idle peer [HOh7S] now active, reconfiguring wireguard
2022/03/10 15:09:03 wgengine: Reconfig: configuring userspace wireguard config (with 2/10 peers)
2022/03/10 15:09:03 magicsock: disco: node [HOh7S] d:c2b10c247333c6cc now using 10.0.5.8:62659
2022/03/10 15:09:03 Accept: UDP{100.x.y.1:6883 > 10.0.5.6:6885} 48 ok
2022/03/10 15:09:03 Accept: UDP{100.x.y.1:6883 > 10.0.5.6:6885} 48 ok
2022/03/10 15:09:03 Accept: UDP{100.x.y.1:6883 > 10.0.5.6:6885} 48 ok
2022/03/10 15:09:03 netstack: could not bind local port 6883: listen udp 0.0.0.0:6883: bind: address already in use, trying again with random port
panic: Incrementing non-positive count 0xc000533180 on stack.PacketBuffer
```

### Are there any recent changes that introduced the issue?

Panic first appeared in 1.22.0.

### OS

macOS

### Tailscale version

1.22

<img src="https://front.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_5u98x)

**Comment by Bleistein** [comment]
Steps to reproduce: On every startup of tailscaled during the first seconds

OS version: 12.2.1 and 11.6.4

**Comment by bradfitz** [comment]
#4207 also had this crash:

```
panic: Incrementing non-positive count 0x400037dc00 on stack.PacketBuffer

goroutine 281 [running]:
gvisor.dev/gvisor/pkg/tcpip/stack.(*packetBufferRefs).IncRef(0x400037dc00)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/packet_buffer_refs.go:80 +0x164
gvisor.dev/gvisor/pkg/tcpip/transport/udp.(*endpoint).HandlePacket(0x4000542000, {0xd9d, {0x400058abc0, 0x4}, 0x4014, {0x400058abbc, 0x4}}, 0x400037dc00)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/transport/udp/endpoint.go:924 +0x444
gvisor.dev/gvisor/pkg/tcpip/stack.(*endpointsByNIC).handlePacket(0x40002da0f0, {0xd9d, {0x400058abc0, 0x4}, 0x4014, {0x400058abbc, 0x4}}, 0x400037dc00)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/transport_demuxer.go:185 +0x2a8
gvisor.dev/gvisor/pkg/tcpip/stack.(*transportDemuxer).deliverPacket(0x400000cc90, 0x11, 0x400037dc00, {0xd9d, {0x400058abc0, 0x4}, 0x4014, {0x400058abbc, 0x4}})
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/transport_demuxer.go:595 +0x414
gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket(0x400024a000, 0x11, 0x400037dc00)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/nic.go:861 +0x310
gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket(0x40000bb100, {0x400003c550, 0x14, 0x50}, 0x4000294000, {0x0, 0x0})
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/network/ipv4/ipv4.go:998 +0x15f0
gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket(0x40000bb100, 0x4000294000)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/network/ipv4/ipv4.go:784 +0x564
gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket(0x400024a000, {0x0, 0x0}, {0x0, 0x0}, 0x800, 0x4000294000)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/nic.go:777 +0x340
gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectLinkAddr(...)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/link/channel/channel.go:196
gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectInbound(...)
	gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/link/channel/channel.go:191
tailscale.com/wgengine/netstack.(*Impl).injectInbound(0x4000154460, 0x4000278680, 0x4000204000)
	tailscale.com/wgengine/netstack/netstack.go:577 +0x66c
tailscale.com/net/tstun.(*Wrapper).filterIn(0x4000204000, {0x40006fc010, 0x47, 0xffef})
	tailscale.com/net/tstun/wrap.go:614 +0x718
tailscale.com/net/tstun.(*Wrapper).Write(0x4000204000, {0x40006fc000, 0x57, 0xffff}, 0x10)
	tailscale.com/net/tstun/wrap.go:627 +0xb4
golang.zx2c4.com/wireguard/device.(*Peer).RoutineSequentialReceiver(0x4000350380)
	golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:477 +0x560
created by golang.zx2c4.com/wireguard/device.(*Peer).Start
	golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/peer.go:199 +0x2bc
```

But it was running linux/arm64 (on 1.22.1)

So I don't think there's anything OS or architecture specific here.

@raggi, as you were just in this code yesterday, maybe it's in your head and you can debug?

**Comment by raggi** [comment]
Yeah, I will take a look, I'm keen to understand where we're managing the buffer refcounts.

**Comment by DentonGentry** [comment]
Does https://github.com/tailscale/tailscale/pull/4220 impact this issue?

**Comment by raggi** [comment]
I don't believe so. I've not fully reproduced this crash yet, but I believe so far that it's specific to exit node configuration.

**Comment by evilhamsterman** [comment]
I'm running into the same thing on 1.21.120 and 1.22.1. I'm running tailscale on Ubuntu 20.04 in kubernetes. It is fine until traffic starts passing through. It only seems to happen on UDP connections. For instance I have both a Minecraft (TCP) and a Valheim (UDP) server running. Minecraft never has an issue but Valheim causes it to crash after a few seconds to a few minutes. 

```
Starting tailscaled
Running tailscale up
2022/04/05 03:46:15 logtail started
2022/04/05 03:46:15 Program starting: v1.21.120-tbb94561c9, Go 1.17.7-tsfd6b85b80f: []string{"tailscaled", "--state=kube:valheim", "--socket=/tmp/tailscaled.sock", "--tun=userspace-networking"}
2022/04/05 03:46:15 LogID: a5ead1aba489f3e5a576db3596a61d14712a81fe38a50c7025dfbedfab277cf8
2022/04/05 03:46:15 logpolicy: using system state directory "/var/lib/tailscale"
logpolicy.Read /var/lib/tailscale/tailscaled.log.conf: open /var/lib/tailscale/tailscaled.log.conf: no such file or directory
2022/04/05 03:46:15 wgengine.NewUserspaceEngine(tun "userspace-networking") ...
2022/04/05 03:46:15 dns: using dns.noopManager
2022/04/05 03:46:15 link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[10.42.0.136/24]} v4=true v6=false}
2022/04/05 03:46:15 magicsock: disco key = d:e3e1f4693a7f1be5
2022/04/05 03:46:15 Creating wireguard device...
2022/04/05 03:46:15 Bringing wireguard device up...
2022/04/05 03:46:15 Bringing router up...
2022/04/05 03:46:15 Clearing router settings...
2022/04/05 03:46:15 Starting link monitor...
2022/04/05 03:46:15 Engine created.
2022/04/05 03:46:15 Start
2022/04/05 03:46:15 using backend prefs for "_daemon": Prefs{ra=false dns=false want=true routes=[] nf=on host="valheim" Persist{lm=, o=, n=[jnndu] u="evilhamsterman@gmail.com"}}
2022/04/05 03:46:15 Backend: logs: be:a5ead1aba489f3e5a576db3596a61d14712a81fe38a50c7025dfbedfab277cf8 fe:
2022/04/05 03:46:15 control: client.Login(false, 0)
2022/04/05 03:46:15 control: doLogin(regen=false, hasUrl=false)
2022/04/05 03:46:15 ipnserver: conn2: connection from userid 0; root has access
2022/04/05 03:46:15 Start
2022/04/05 03:46:15 control: client.Shutdown()
2022/04/05 03:46:15 control: client.Shutdown: inSendStatus=0
2022/04/05 03:46:15 control: mapRoutine: quit
2022/04/05 03:46:15 control: Client.Shutdown done.
2022/04/05 03:46:15 using backend prefs for "_daemon": Prefs{ra=false dns=false want=true routes=[] nf=on host="valheim" Persist{lm=, o=, n=[jnndu] u="evilhamsterman@gmail.com"}}
2022/04/05 03:46:15 Backend: logs: be:a5ead1aba489f3e5a576db3596a61d14712a81fe38a50c7025dfbedfab277cf8 fe:
2022/04/05 03:46:15 control: client.Login(false, 0)
2022/04/05 03:46:15 control: doLogin(regen=false, hasUrl=false)
2022/04/05 03:46:16 control: control server key [nlFWp] from https://controlplane.tailscale.com
2022/04/05 03:46:16 control: RegisterReq: onode= node=[jnndu] fup=false
2022/04/05 03:46:16 control: RegisterReq: got response; nodeKeyExpired=false, machineAuthorized=true; authURL=false
2022/04/05 03:46:16 active login: evilhamsterman@gmail.com
2022/04/05 03:46:16 Switching ipn state NoState -> Starting (WantRunning=true, nm=true)
2022/04/05 03:46:16 health("overall"): error: state=Starting, wantRunning=true
2022/04/05 03:46:16 magicsock: SetPrivateKey called (init)
2022/04/05 03:46:16 wgengine: Reconfig: configuring userspace wireguard config (with 0/6 peers)
2022/04/05 03:46:16 wgengine: Reconfig: configuring router
2022/04/05 03:46:16 wgengine: Reconfig: configuring DNS
2022/04/05 03:46:16 dns: Set: {DefaultResolvers:[] Routes:{} SearchDomains:[] Hosts:7}
2022/04/05 03:46:16 dns: Resolvercfg: {Routes:{} Hosts:7 LocalDomains:[]}
2022/04/05 03:46:16 dns: OScfg: {Nameservers:[] SearchDomains:[] MatchDomains:[]}
2022/04/05 03:46:16 Taildrop disabled; no state directory
2022/04/05 03:46:16 peerapi starting without Taildrop directory configured
2022/04/05 03:46:16 peerapi: serving on http://100.122.156.102:46575
2022/04/05 03:46:16 peerapi: serving on http://[fd7a:115c:a1e0:ab12:4843:cd96:627a:9c66]:46575
2022/04/05 03:46:17 magicsock: home is now derp-10 (sea)
2022/04/05 03:46:17 magicsock: adding connection to derp-10 for home-keep-alive
2022/04/05 03:46:17 control: NetInfo: NetInfo{varies=true hairpin=false ipv6=false udp=true derp=#10 portmap= link=""}
2022/04/05 03:46:17 magicsock: 1 active derp conns: derp-10=cr0s,wr0s
2022/04/05 03:46:17 derphttp.Client.Connect: connecting to derp-10 (sea)
2022/04/05 03:46:17 magicsock: endpoints changed: 24.113.14.58:5149 (stun), 10.42.0.136:57850 (local)
2022/04/05 03:46:17 Switching ipn state Starting -> Running (WantRunning=true, nm=true)
2022/04/05 03:46:17 ipnserver: conn2: ReadMsg: read unix /tmp/tailscaled.sock->@: read: connection reset by peer
2022/04/05 03:46:17 magicsock: derp-10 connected; connGen=1
2022/04/05 03:46:17 health("overall"): ok
2022/04/05 03:47:57 wgengine: idle peer [BrnUG] now active, reconfiguring wireguard
2022/04/05 03:47:57 wgengine: Reconfig: configuring userspace wireguard config (with 1/6 peers)
2022/04/05 03:47:57 magicsock: disco: node [BrnUG] d:43f4516aa6a2df20 now using 192.168.1.126:41641
2022/04/05 03:47:57 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 540 ok
2022/04/05 03:47:57 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 360 ok
2022/04/05 03:47:57 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 70 ok
2022/04/05 03:48:07 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 58 ok
2022/04/05 03:48:17 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 1202 ok
2022/04/05 03:48:27 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 71 ok
2022/04/05 03:48:37 Accept: UDP{100.94.66.37:59573 > 100.122.156.102:2456} 978 ok
panic: Incrementing non-positive count 0xc0000ea380 on stack.PacketBuffer
goroutine 372 [running]:
gvisor.dev/gvisor/pkg/tcpip/stack.(*packetBufferRefs).IncRef(0xc0000ea380)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/packet_buffer_refs.go:80 +0x105
gvisor.dev/gvisor/pkg/tcpip/transport/udp.(*endpoint).HandlePacket(0xc0001c4c00, {0x998, {0xc00055a7d0, 0x4}, 0xe8b5, {0xc00055a7cc, 0x4}}, 0xc0000ea380)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/transport/udp/endpoint.go:924 +0x2ed
gvisor.dev/gvisor/pkg/tcpip/stack.(*endpointsByNIC).handlePacket(0xc0006331d0, {0x998, {0xc00055a7d0, 0x4}, 0xe8b5, {0xc00055a7cc, 0x4}}, 0xc0000ea380)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/transport_demuxer.go:185 +0x27f
gvisor.dev/gvisor/pkg/tcpip/stack.(*transportDemuxer).deliverPacket(0xc0004601c8, 0x11, 0xc0000ea380, {0x998, {0xc00055a7d0, 0x4}, 0xe8b5, {0xc00055a7cc, 0x4}})
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/transport_demuxer.go:595 +0x3c5
gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket(0xc00000a1e0, 0x11, 0xc0000ea380)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/nic.go:861 +0x1fe
gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket(0xc0002a3800, {0xc00028c160, 0x14, 0x20}, 0xc0000ea700, {0x0, 0x0})
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/network/ipv4/ipv4.go:998 +0xfd6
gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket(0xc0002a3800, 0x886375)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/network/ipv4/ipv4.go:784 +0x2b8
gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket(0xc00000a1e0, {0xc0007efbf0, 0x1}, {0x0, 0x0}, 0x800, 0xc0000ea700)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/nic.go:777 +0x2dd
gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectLinkAddr(...)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/link/channel/channel.go:196
gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectInbound(...)
gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/link/channel/channel.go:191
tailscale.com/wgengine/netstack.(*Impl).injectInbound(0xc0001e0000, 0xc0000a0080, 0xc0000aad80)
tailscale.com/wgengine/netstack/netstack.go:577 +0x744
tailscale.com/net/tstun.(*Wrapper).filterIn(0xc000214000, {0xc0000b8010, 0x1b, 0xffef})
tailscale.com/net/tstun/wrap.go:614 +0x78d
tailscale.com/net/tstun.(*Wrapper).Write(0xc000214000, {0xc0000b8000, 0xc0002e2a00, 0x1}, 0x10)
tailscale.com/net/tstun/wrap.go:627 +0x7c
golang.zx2c4.com/wireguard/device.(*Peer).RoutineSequentialReceiver(0xc000595c00)
golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:477 +0x4d1
created by golang.zx2c4.com/wireguard/device.(*Peer).Start
golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/peer.go:199 +0x295```

@DentonGentry were you running UDP traffic or TCP when it crashed. 

My uneducated guess is since UDP is stateless it's constantly creating new sessions and there's a race condition somewhere

**Comment by raggi** [comment]
I've been trying to reproduce this issue for a while, so far I have not found a way to do so. I had previously attempted to reproduce the issue using iperf3 and some hand-rolled udp bind spam to try to mimic what I've seen in some of the user reports.

This evening I setup a valheim server and a couple of us played for a couple of hours. Everything was stable for us, on ubuntu 20.04 at 1.22.2 / tailscale commit: 60b67195554487df2b40d49ad8bfc77f8c734b4e. It was fun, and did plenty of traffic  tx 168224304 rx 71748880 / tx 71807684 rx 168225932.

![image](https://user-images.githubusercontent.com/348/161884458-ff0a0922-df7c-4382-905b-c4e1de4cbe30.png)

A fun, but sadly not so fruitful debugging session.

I also tested with 1.21.1, and v1.21.120-tbb94561c9, Go 1.17.7-tsfd6b85b. I don't recognize the Go version you're using, it does not appear to be what build_dist.sh pulls for bb94561c9. I've not so far been able to reproduce the crash.

How many cores does this container have access to?

**Comment by merrickluo** [comment]
This happens **a lot** to me when running a exit node on a TrueNAS(FreeBSD) jail, the stack trace looks slightly different though
```
goroutine 30908 [running]:
gvisor.dev/gvisor/pkg/tcpip/stack.(*packetBufferRefs).IncRef(0xc000799c00)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/packet_buffer_refs.go:80 +0x16a
gvisor.dev/gvisor/pkg/tcpip/transport/udp.(*endpoint).HandlePacket(0xc000def500, {0x1bb, {0xc000284b58, 0x4}, 0xc791, {0xc000284b54, 0x4}}, 0xc000799c00)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/transport/udp/endpoint.go:924 +0x305
gvisor.dev/gvisor/pkg/tcpip/transport/udp.(*ForwarderRequest).CreateEndpoint(0xc000554200, 0xc0004a22a0)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/transport/udp/forwarder.go:100 +0x45a
tailscale.com/wgengine/netstack.(*Impl).acceptUDP(0xc00015c640, 0xc000554200)
        tailscale.com/wgengine/netstack/netstack.go:730 +0x1b9
gvisor.dev/gvisor/pkg/tcpip/transport/udp.(*Forwarder).HandlePacket(...)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/transport/udp/forwarder.go:47
tailscale.com/wgengine/netstack.(*Impl).wrapProtoHandler.func1({0x1bb, {0xc000284b58, 0x4}, 0xc791, {0xc000284b54, 0x4}}, 0xc000799c00)
        tailscale.com/wgengine/netstack/netstack.go:198 +0x148
gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverTransportPacket(0xc00000c1e0, 0x11, 0xc000799c00)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/nic.go:867 +0x2f1
gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).handleValidatedPacket(0xc0000bdc00, {0xc0004ae100, 0x14, 0x80}, 0xc000726000, {0x0, 0x0})
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/network/ipv4/ipv4.go:998 +0x1233
gvisor.dev/gvisor/pkg/tcpip/network/ipv4.(*endpoint).HandlePacket(0xc0000bdc00, 0xc000726000)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/network/ipv4/ipv4.go:784 +0x39b
gvisor.dev/gvisor/pkg/tcpip/stack.(*nic).DeliverNetworkPacket(0xc00000c1e0, {0x0, 0x0}, {0x0, 0x0}, 0x800, 0xc000726000)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/stack/nic.go:777 +0x349
gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectLinkAddr(...)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/link/channel/channel.go:196
gvisor.dev/gvisor/pkg/tcpip/link/channel.(*Endpoint).InjectInbound(...)
        gvisor.dev/gvisor@v0.0.0-20220126021142-d8aa030b2591/pkg/tcpip/link/channel/channel.go:191
tailscale.com/wgengine/netstack.(*Impl).injectInbound(0xc00015c640, 0xc00091e000, 0xc0002a8000)
        tailscale.com/wgengine/netstack/netstack.go:577 +0x765
tailscale.com/net/tstun.(*Wrapper).filterIn(0xc0002a8000, {0xc0010e0010, 0x7a, 0xffef})
        tailscale.com/net/tstun/wrap.go:614 +0x893
tailscale.com/net/tstun.(*Wrapper).Write(0xc0002a8000, {0xc0010e0000, 0x8a, 0xffff}, 0x10)
        tailscale.com/net/tstun/wrap.go:627 +0x90
golang.zx2c4.com/wireguard/device.(*Peer).RoutineSequentialReceiver(0xc000266700)
        golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:477 +0x5b7
created by golang.zx2c4.com/wireguard/device.(*Peer).Start
        golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/peer.go:199 +0x2d1
```

it happened 5+ times today, and it seems more likely to happen when a new device joins, but not always.

**Comment by evilhamsterman** [comment]
> I've been trying to reproduce this issue for a while, so far I have not found a way to do so. I had previously attempted to reproduce the issue using iperf3 and some hand-rolled udp bind spam to try to mimic what I've seen in some of the user reports.
> 
> This evening I setup a valheim server and a couple of us played for a couple of hours. Everything was stable for us, on ubuntu 20.04 at 1.22.2 / tailscale commit: 60b67195554487df2b40d49ad8bfc77f8c734b4e. It was fun, and did plenty of traffic tx 168224304 rx 71748880 / tx 71807684 rx 168225932.
> 
> ![image](https://user-images.githubusercontent.com/348/161884458-ff0a0922-df7c-4382-905b-c4e1de4cbe30.png)
> 
> A fun, but sadly not so fruitful debugging session.
> 
> I also tested with 1.21.1, and v1.21.120-tbb94561c9, Go 1.17.7-tsfd6b85b. I don't recognize the Go version you're using, it does not appear to be what build_dist.sh pulls for [bb94561](https://github.com/tailscale/tailscale/commit/bb94561c96a4a8f4089acadbdda58aa8f0e83067). I've not so far been able to reproduce the crash.
> 
> How many cores does this container have access to?

I'm glad I gave you and your team an excuse to play Valheim.

It is running on a cluster of mini systems with AMD Ryzen 5 PRO 2400GE processors 4 cores/8 threads. The container has no resource limits set so it can use the full resources of the machine and is currently the only thing of consequence on the system. I'm using k3s stable release (1.22) as my K8S distro.

I'm using a derivative of the tailscale/tailscale container from Docker Hub where I added the run.sh script that I pulled from your K8S docs, the only thing I changed was added support for setting the Tailscale hostname. It is being deployed as a sidecar container to my Valheim and Minecraft pods. The Dockerfile and run.sh are in my repo here https://gitlab.com/evilhamsterman/tailscale-k8s

I've pushed my container to Docker hub for both the 1.20, 1.21, and 1.22 versions. I haven't had a chance to see if the same thing happens in 1.20

**Comment by evilhamsterman** [comment]
I played for about 10-15 minutes with a 1.20 version of tailscale as the sidecar and had no issues, usually it would crash within a minute or so.

**Comment by evilhamsterman** [comment]
One more datapoint. I was initially using tailscale in userspace mode. I switched to normal mode on 1.22 and the problem no longer occurs. So it appears to happen with UDP in version 1.21 and 1.22 when running in userspace mode, which would explain why it also happens on MacOS

**Comment by raggi** [comment]
Thank you for the detailed information. I setup a k3s cluster today, created a pod in much the same way.

```
2022/04/07 00:35:09 Program starting: v1.21.120-tbb94561c9, Go 1.17.7-tsfd6b85b80f: []string{"tailscaled", "--state=kube:tailscale-auth", "--socket=/tmp/tailscaled.sock", "--tun=userspace-networking"}
```

alas, still no reproduction.

The code in question pushes packetbufs through a sync.pool, which is cleared on GC cranks, and there is at least one bug on the code path though it does not appear to be caused in a normal configuration by this setup (here: https://github.com/google/gvisor/pull/7336). It is however possible that GC rate affects the likelihood of a reproduction of this issue. How much ram is your tailscaled instance using?

**Comment by jonseymour** [comment]
We are also experiencing this issue on both 1.22.0 and 1.22.2. 

I wonder if the switch from inet.af/netstack to gvisor.dev/gvisor in this commit  -> 1af26222 introduced the instability by dragging an undesirable upstream change?

**Comment by raggi** [comment]
If you are able to try 1.23.144 or later, the gvisor roll there may address this issue. There are packages available for most builds on https://pkgs.tailscale.com/unstable. There is also an unstable tag with 1.23.148 at ghcr.io/tailscale/tailscale:unstable.

@jonseymour thank you. I believe that the PreserveObject API that was in upstream in this range is dangerous in ways that could lead to the reported crash, and that is now removed in the netstack roll that is included after 1.23.112 (73314009d0a600a3b0b6aced4c9d9c65d89ffbb7 / https://github.com/google/gvisor/commit/f375784d83852b1e3ff20cc9de0648b3c0cf8525), which also adds more complete refcounting to packetbuf paths in the tun drivers. I did read through the code a few times looking for a code path that could result in a packetbuf being added to the pool that has been preserved and deref'd, or that has been deref'd and passed on - I did not spot one (though clearly there is a bug somewhere). I'm keen to find a reproduction so that I can add regression tests for this case.

**Comment by evilhamsterman** [comment]
@raggi I will give it a go as soon as I can, going to be busy the next few days.

**Comment by evilhamsterman** [comment]
I played for a little bit at lunch time using 1.23.148. No issues occurred

**Comment by raggi** [comment]
Ok, reproduced.

The crash is caused by fragment handling in the userspace networking path. The easiest method to reproduce the issue is to use a normal mode configuration to send a larger than host mtu packet to a node in userspace-networking.

Further digging in gvisor reveals it's fixed in https://github.com/google/gvisor/commit/6a28dc7c59632b4007a095377073b8b74df85bea so this issue will be addressed with the 1.24 release.

**Comment by jonseymour** [comment]
@evilhamsterman v1.20.4 doesn't have this issue, but there does appear to be another issue  which is causing the tailscale daemon to quit (but not panic). I haven't fully characterised this issue yet, so it may be a known issue.

FWIW: it seems if the comments by the gvisor developer are correct, the unreleased commit 73314009 (#4222) would have been sufficient to address this issue, which is not to say that updating to a later version isn't the right thing to do.

**Comment by raggi** [comment]
@jonseymour very happy to discuss that issue on a new bug. I expect that the bulk of the issue is already fixed in 1.24, as one of the main reasons I initiated a bump of netstack for 1.24 is for improved buffer recycling, and I also did some work on our side to make better use of the buffer pools as well, in userspace mode.

The other detail in your report which is the udp session timeout could potentially lead to an issue if there was enough instances of new sessions being started (several orders of magnitude). Please go ahead and open a bug about that, and if you are able to provide a little information about the traffic patterns producing the scenario that would be helpful.

**Comment by jonseymour** [comment]
Thanks - done - #4381
